### PR TITLE
Add linting and formating

### PR DESCRIPTION
**Pull request by JbIPS**
To enable a better collaboration, we should enforce Ruff linting and Black formatting.

* Added Ruff and Black in requirements
* Created a GitHub workflow to check linting and formatting on api push

### CODE REVIEW

**Comment by JbIPS** [comment]
I took the liberty to implement some basic housekeeping and I allowed me to have a closer look to the code.

I think I'll soon create an issue about the project architecture/scaffolding :wink:

File: api/activetigger/functions.py [hunk 1]
```diff
@@ -41,10 +42,40 @@
...
+        progress_percentage = (state.global_step / state.max_steps) * 100
+        with open(self.current_path / "train/progress", "w") as f:
+            f.write(str(progress_percentage))
+        # end if event set
+        if self.event is not None:
+            if self.event.is_set():
+                self.logger.info("Event set, stopping training.")
+                control.should_training_stop = True
+
+
```
Comment (Carreau) [inline api/activetigger/functions.py:78]: 👍

File: api/activetigger/functions.py [hunk 2]
```diff
@@ -41,10 +42,40 @@
...
+    def on_step_end(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        self.logger.info(f"Step {state.global_step}")
+        progress_percentage = (state.global_step / state.max_steps) * 100
+        with open(self.current_path / "train/progress", "w") as f:
```
Comment (Carreau) [inline api/activetigger/functions.py:70]: I'm quasi sure that if `current_path` is a `str`, then `self.current_path / "train/progress"` will fail on the `/`, it is likely `current_path` is a `pathlib.Path`, 

And you want to do `(self.curent_path/"train"/"progress").write_text(data)` (also I'm not sure of the cross-platform behavior of having a `/`  in a string; it is possible, that `self.current_path / "train/progress"`  could create a file actually named 'train/progress'; but I think you want the `/` outside the string.

(I know the `/` things is from the original version ).


Also to catch these types of errors; later on you want to turn on [mypy](https://pypi.org/project/mypy/) in CI.

File: api/activetigger/functions.py [hunk 3]
```diff
@@ -738,8 +760,8 @@ def request_ollama(endpoint: str, request: str, model: str = "llama3.1:70b"):
     if response.status_code == 200:
         try:
             return {"success": response.json()["response"]}
-        except:
-            return {"error": "Error in the content"}
+        except Exception as e:
+            return {"error": f"Error in the content: {e}"}
```
Comment (Carreau) [inline api/activetigger/functions.py:764]: +1, 

Be mindful the traceback may contain sensitive information if verbose traceback are activate (e.g if a local variable contain an API Key, verbose tb could contain the value of the local variable.

**Review by Carreau — APPROVED** [review]

File: api/activetigger/functions.py (hunk 3 above)
Comment (emilienschultz) [reply api/activetigger/functions.py:768]: Ok, i see the point

File: api/activetigger/functions.py [hunk 4]
```diff
@@ -552,20 +577,21 @@ def on_step_end(self, args, state, control, **kwargs):
...
         # save training data
-        training_data.to_parquet(current_path / "training_data.parquet")
+        training_data.to_parquet(
+            current_path.joinpath("training_data.parquet"))
 
         # save parameters
         params["test_size"] = test_size
         params["base_model"] = base_model
-        with open(current_path / "parameters.json", "w") as f:
+        with open(current_path.joinpath("parameters.json", "w")) as f:
```
Comment (emilienschultz) [inline api/activetigger/functions.py:586]: There is a problem here in the parenthesis

**Review by emilienschultz — APPROVED** [review]

**Comment by emilienschultz** [comment]
There is one mistake in "with open(current_path.joinpath("parameters.json", "w")) as f:" but i will fix it

**Pull request merged**